Storing the address of a local variable

### Game and System Information

- **Version:** 1.19.11+dev
- **OS:** Debian 12.10


### Description of the bug

> g++ -o build/release/tooltips.o -c -Werror=non-virtual-dtor -std=c++17 -Wold-style-cast -Wall -Wextra -pthread -Wno-unused-local-typedefs -Wno-maybe-uninitialized -Wtrampolin
es -fPIE -fstack-protector-strong -O3 -D_REENTRANT -DHAVE_LIBDBUS -DHAVE_HISTORY -DHAVE_CONFIG_H -D_FORTIFY_SOURCE=2 -D_X11 -DFIFODIR='"/var/run/wesnothd"' -DWESNOTH_PATH='"/
usr/local/share/wesnoth"' -DLOCALEDIR='"translations"' -DHAS_RELATIVE_LOCALEDIR -Ibuild -I/usr/src/wesnoth/master/build -Isrc -I/usr/src/wesnoth/master/src -isystem /usr/include/SDL2 -isystem /usr/include/cairo -isystem /usr/include/glib-2.0 -isystem /usr/lib/x86_64-linux-gnu/glib-2.0/include -isystem /usr/include/pixman-1 -isystem /usr/include/freetype2 -isystem /usr/include/libpng16 -isystem /usr/include/pango-1.0 -isystem /usr/include/harfbuzz -isystem /usr/include/libmount -isystem /usr/include/blkid -isystem /usr/include/fribidi -isystem /usr/include/dbus-1.0 -isystem /usr/lib/x86_64-linux-gnu/dbus-1.0/include /usr/src/wesnoth/master/src/tooltips.cpp
In file included from /usr/include/c++/12/set:60,
                 from /usr/src/wesnoth/master/src/map/location.hpp:23,
                 from /usr/src/wesnoth/master/src/terrain/builder.hpp:24,
                 from /usr/src/wesnoth/master/src/terrain/builder.cpp:21:
In member function ‘void std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::swap(std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>&) [with _Key = terrain_builder::building_rule; _Val = terrain_builder::building_rule; _KeyOfValue = std::_Identity<terrain_builder::building_rule>; _Compare = std::less<terrain_builder::building_rule>; _Alloc = std::allocator<terrain_builder::building_rule>]’,
    inlined from ‘void std::multiset<_Key, _Compare, _Alloc>::swap(std::multiset<_Key, _Compare, _Alloc>&) [with _Key = terrain_builder::building_rule; _Compare = std::less<terrain_builder::building_rule>; _Alloc = std::allocator<terrain_builder::building_rule>]’ at /usr/include/c++/12/bits/stl_multiset.h:440:18,
    inlined from ‘void std::swap(multiset<_Key, _Compare, _Allocator>&, multiset<_Key, _Compare, _Allocator>&) [with _Key = terrain_builder::building_rule; _Compare = less<terrain_builder::building_rule>; _Alloc = allocator<terrain_builder::building_rule>]’ at /usr/include/c++/12/bits/stl_multiset.h:1049:15,
    inlined from ‘static void terrain_builder::set_terrain_rules_cfg(const game_config_view&)’ at /usr/src/wesnoth/master/src/terrain/builder.cpp:285:11:
/usr/include/c++/12/bits/stl_tree.h:2091:36: warning: storing the address of local variable ‘empty’ in ‘*MEM[(struct _Rb_tree_node_base * &)&empty + 16].std::_Rb_tree_node_base::_M_parent’ [-Wdangling-pointer=]
 2091 |           __t._M_root()->_M_parent = __t._M_end();
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/usr/src/wesnoth/master/src/terrain/builder.cpp: In static member function ‘static void terrain_builder::set_terrain_rules_cfg(const game_config_view&)’:
/usr/src/wesnoth/master/src/terrain/builder.cpp:284:26: note: ‘empty’ declared here
  284 |         building_ruleset empty;
      |                          ^~~~~
/usr/src/wesnoth/master/src/terrain/builder.cpp:284:26: note: ‘empty.std::multiset<terrain_builder::building_rule, std::less<terrain_builder::building_rule>, std::allocator<terrain_builder::building_rule> >::_M_t.std::_Rb_tree<terrain_builder::building_rule, terrain_builder::building_rule, std::_Identity<terrain_builder::building_rule>, std::less<terrain_builder::building_rule>, std::allocator<terrain_builder::building_rule> >::_M_impl.std::_Rb_tree<terrain_builder::building_rule, terrain_builder::building_rule, std::_Identity<terrain_builder::building_rule>, std::less<terrain_builder::building_rule>, std::allocator<terrain_builder::building_rule> >::_Rb_tree_impl<std::less<terrain_builder::building_rule>, true>::<unnamed>.std::_Rb_tree_header::_M_header.std::_Rb_tree_node_base::_M_parent’ declared here

https://github.com/wesnoth/wesnoth/blob/38d163ff8eba1556ff1bc76fc77779eefef1d122/src/terrain/builder.cpp#L284-L285


### Steps to reproduce the behavior

Pretty much a default build of master with scons on Debian 12.10.

### Expected behavior

No compiler warnings.

### Additional context

The code is not new (https://github.com/wesnoth/wesnoth/commit/e3b61579f0aebded6224be56f46d237fab2df84b) but it seems like it's undefined behavior and we got lucky so far?

--- a/src/terrain/builder.cpp
+++ b/src/terrain/builder.cpp
@@ -279,10 +279,7 @@ void terrain_builder::flush_local_rules()
 void terrain_builder::set_terrain_rules_cfg(const game_config_view& cfg)
 {
 	rules_cfg_ = &cfg;
-	// use the swap trick to clear the rules cache and get a fresh one.
-	// because simple clear() seems to cause some progressive memory degradation.
-	building_ruleset empty;
-	std::swap(building_rules_, empty);
+	building_rules_.clear();
 }
 
 void terrain_builder::reload_map()

